source/ui/pixel-display-painted-item.cpp

```diff
@@ -0,0 +1,738 @@
+        [](const MemoryValue &buffer,
+           std::size_t offset,
+           const Options &o,
+           std::size_t index) -> std::uint32_t {
+  std::size_t cellSize = 8;// TODO
+  constexpr std::size_t colorSizeInBit = 32;
+  std::size_t colorSize = (colorSizeInBit + cellSize - 1) / cellSize;
+  std::size_t byteOffset = (colorSize * index) - offset;
+  if (buffer.getSize() >= byteOffset * cellSize + colorSizeInBit) {
+    MemoryValue color = buffer.subSet(byteOffset * cellSize,
+                                      byteOffset * cellSize + colorSizeInBit);
+    return conversions::convert<std::uint32_t>(
+        color, conversions::standardConversions::nonsigned);
+  }
+  return 0xFFFF00FF;// error color
+};
+
+const ColorMode::UpdateMemoryFunction ColorMode::MonochromeUpdateMemory = [](
+    Optional<OutputComponent *> memoryAccess,
+    const Options &o,
+    std::shared_ptr<QImage> image,
+    std::size_t address,
+    std::size_t amount) -> void {
+  std::size_t cellSize = 8;    // TODO
+  std::size_t pointerSize = 32;// TODO
+  std::size_t pointerSizeInByte = (pointerSize + cellSize - 1) / cellSize;
+  std::size_t pixelBufferPointer = o.pixelBaseAddress;
+  bool hasAlreadyUpdatedAllPixels = false;
+  if (o.pixelBufferPointerLike) {
+    if (address < pixelBufferPointer + pointerSizeInByte &&
+        address + amount > pixelBufferPointer) {
+      // pixel pointer has changed
+      o.updateAllPixels(memoryAccess, image);
+      hasAlreadyUpdatedAllPixels = true;
+    } else {
+      pixelBufferPointer = ColorMode::loadPointer(memoryAccess,
+                                                  o.pixelBaseAddress,
+                                                  o.pixelBufferPointerLike,
+                                                  cellSize,
+                                                  pointerSize);
+    }
+  }
+  std::size_t pixelBufferSize = (o.width * o.height + cellSize - 1) / cellSize;
+  if (!hasAlreadyUpdatedAllPixels) {
+    if (address < pixelBufferPointer + pixelBufferSize &&
+        address + amount > pixelBufferPointer) {
+      // At least some pixel have changed
+      std::size_t bitsPerByte = cellSize - o.freeBits;
+      std::size_t minIndex = 0;
+      if (address > pixelBufferPointer) {
+        minIndex = (address - pixelBufferPointer) * bitsPerByte;
+      }
+      std::size_t maxIndex = o.height * o.width;
+      if (address + amount < pixelBufferPointer + pixelBufferSize) {
+        maxIndex = (address + amount - pixelBufferPointer) * bitsPerByte;
+      }
+      std::size_t beginOffset = minIndex / bitsPerByte;
+      std::size_t endOffset = (maxIndex + bitsPerByte - 1) / bitsPerByte;
+      MemoryValue buffer = ColorMode::getMemoryValueAt(
+          memoryAccess, pixelBufferPointer, pixelBufferSize);
+      // this could be optimized, we don't actually need the whole pixel buffer,
+      // but anyways... (I'm not sure I'm calculating the offset correctly TODO)
```

Comment:
I am currently not calculating any offset, as I fetch the whole color buffer because the fetching of only parts of the image buffer did not work and I wasn't sure about the formula I used to calculate the offset